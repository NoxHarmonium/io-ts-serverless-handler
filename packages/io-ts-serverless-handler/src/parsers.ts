import { JsonValue } from "type-fest";
import { Parser } from ".";

/**
 * A simple JSON parser.
 *
 * Defined here so it can be overriden for tests,
 * and possibly made pluggable in the future to support
 * different parsers.
 */
export const jsonParser: Parser<JsonValue> = {
  parse: JSON.parse,
};
