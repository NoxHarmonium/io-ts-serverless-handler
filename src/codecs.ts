import { either } from "fp-ts/lib/Either";
import * as t from "io-ts";
import { JsonValue } from "type-fest";

/**
 * A codec that takes a JSON string and does two things:
 *   1. Validates that the input is a string
 *   2. Validates that the string represents a valid JSON object
 * Useful to parse a JSON object that can then be passed into a higher order codec
 */
export const jsonFromStringCodec = new t.Type<JsonValue, string, unknown>(
  "JSONFromString",
  (u): u is JsonValue => u !== undefined,
  (u, c) =>
    either.chain(t.string.validate(u, c), s => {
      try {
        return t.success(JSON.parse(s));
      } catch (e) {
        if (e instanceof Error) {
          return t.failure(u, c, e.message);
        } else {
          return t.failure(u, c, "Unknown error");
        }
      }
    }),
  a => JSON.stringify(a)
);
