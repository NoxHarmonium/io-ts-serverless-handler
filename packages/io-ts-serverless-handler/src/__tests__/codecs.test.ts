import { isLeft, isRight } from "fp-ts/lib/Either";
import each from "jest-each";
import { JsonValue } from "type-fest";
import { jsonFromStringCodec } from "../codecs";
import { jsonParser } from "../parsers";

jest.mock("../parsers");

describe("jsonFromStringCodec", () => {
  beforeEach(() => {
    // Regular implementation
    ((jsonParser.parse as unknown) as jest.Mock).mockImplementation(JSON.parse);
  });

  describe("when decoding a invalid input", () => {
    each([
      [undefined],
      ["undefined"],
      [null],
      [8],
      [false],
      [""],
      ["dfsjlsf"],
      ['{ "hello": 4 '],
    ]).test("fails when decoding [%p]", (input: unknown) => {
      expect.assertions(2);
      const result = jsonFromStringCodec.decode(input);
      expect(isLeft(result)).toBe(true);
      if (isLeft(result)) {
        expect(result.left).toMatchSnapshot();
      }
    });
  });
  describe("when decoding a valid input", () => {
    each([["null"], ["8"], ["false"], ['{ "hello": 4 }']]).test(
      "succeeds when decoding [%p]",
      (input: unknown) => {
        expect.assertions(2);

        const result = jsonFromStringCodec.decode(input);
        expect(isRight(result)).toBe(true);
        if (isRight(result)) {
          expect(result.right).toMatchSnapshot();
        }
      }
    );
  });
  describe("when encoding a JsonValue", () => {
    each([[null], [8], [false], [{ hello: 4 }]]).test(
      "succeeds when encoding [%p]",
      (input: JsonValue) => {
        expect.assertions(2);

        const isJsonValue = jsonFromStringCodec.is(jsonFromStringCodec);
        expect(isJsonValue).toBe(true);
        const result = jsonFromStringCodec.encode(input);
        expect(result).toMatchSnapshot();
      }
    );
  });
  describe("when an error that does not inherit from the Error object is thrown", () => {
    it("should propagate it and not crash", () => {
      expect.assertions(2);

      ((jsonParser.parse as unknown) as jest.Mock).mockImplementation(() => {
        // tslint:disable-next-line: no-string-throw
        throw "something that is not an Error object";
      });

      const result = jsonFromStringCodec.decode("*DFHDF7890");
      expect(isLeft(result)).toBe(true);
      if (isLeft(result)) {
        expect(result.left).toMatchSnapshot();
      }
    });
  });
});
