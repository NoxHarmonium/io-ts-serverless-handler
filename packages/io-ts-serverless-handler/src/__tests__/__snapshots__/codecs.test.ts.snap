// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonFromStringCodec when an error that does not inherit from the Error object is thrown should propagate it and not crash 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": "*DFHDF7890",
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": "Unknown error",
    "value": "*DFHDF7890",
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding [""] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": "",
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": "Unexpected end of JSON input",
    "value": "",
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding ["{ \\"hello\\": 4 "] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": "{ \\"hello\\": 4 ",
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": "Unexpected end of JSON input",
    "value": "{ \\"hello\\": 4 ",
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding ["dfsjlsf"] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": "dfsjlsf",
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": "Unexpected token d in JSON at position 0",
    "value": "dfsjlsf",
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding ["undefined"] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": "undefined",
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": "Unexpected token u in JSON at position 0",
    "value": "undefined",
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding [8] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": 8,
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": 8,
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding [false] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": false,
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": false,
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding [null] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": null,
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": null,
  },
]
`;

exports[`jsonFromStringCodec when decoding a invalid input fails when decoding [undefined] 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": undefined,
        "key": "",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "JSONFromString",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": undefined,
  },
]
`;

exports[`jsonFromStringCodec when decoding a valid input succeeds when decoding ["{ \\"hello\\": 4 }"] 1`] = `
Object {
  "hello": 4,
}
`;

exports[`jsonFromStringCodec when decoding a valid input succeeds when decoding ["8"] 1`] = `8`;

exports[`jsonFromStringCodec when decoding a valid input succeeds when decoding ["false"] 1`] = `false`;

exports[`jsonFromStringCodec when decoding a valid input succeeds when decoding ["null"] 1`] = `null`;

exports[`jsonFromStringCodec when encoding a JsonValue succeeds when encoding [{"hello": 4}] 1`] = `"{\\"hello\\":4}"`;

exports[`jsonFromStringCodec when encoding a JsonValue succeeds when encoding [8] 1`] = `"8"`;

exports[`jsonFromStringCodec when encoding a JsonValue succeeds when encoding [false] 1`] = `"false"`;

exports[`jsonFromStringCodec when encoding a JsonValue succeeds when encoding [null] 1`] = `"null"`;
